cpp/ql/src/Security/CWE/CWE-732/OpenCallMissingModeArgument.ql

```diff
@@ -0,0 +1,19 @@
+/**
+ * @name File opened with O_CREAT flag but without mode argument
+ * @description Opening a file with the O_CREAT flag but without mode argument reads arbitrary bytes from the stack.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.8
+ * @precision medium
+ * @id cpp/open-call-with-mode-argument
+ * @tags security
+ *       external/cwe/cwe-732
```

Comment:
`precision` and `problem.severity`: are very broadly described in https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md

In my own words, precision describes the proportion of results a user should expect to be correct (not just technically correct according to the specification of the query, which may itself have shortcomings).  `medium` suggests a balance of good results and false positives, but not overwhelmingly false positives.  `high` suggests that most results will be correct.  `very-high` suggests you are unlikely to find any false positive results at all (we rarely use `very-high` in CPP because its a complex language with very varied practices).

The other dimension, severity, is about how bad a true positive result is for the project.  `error` means the result is likely to indicate incorrect program behaviour or a security vulnerability, whereas a `warning` or `recommendation` might have less severe / immediate consequences.

The default code scanning suite I believe only shows queries with the `security` tag (I think), a `precision` of `high` or `very-high` and a severity of `warning` or `error` (this is a very high standard).  The security-extended suite is the same but also shows `medium` precision queries, so users expect to get more results at the cost of sorting through more false positives.

Security-severity: there's a (internal) script to generate these.

Reply:
> The default code scanning suite I believe only shows queries with the security tag (I think), a precision of high or very-high and a severity of warning or error (this is a very high standard). The security-extended suite is the same but also shows medium precision queries, so users expect to get more results at the cost of sorting through more false positives.

This is mostly correct (the security-extended suite doesn't show the medium precision queries by default, though). Here's how you can find the formalization of these rules:

Code Scanning uses the `.qls` file from [here](https://github.com/github/codeql/blob/main/cpp/ql/src/codeql-suites/cpp-code-scanning.qls). That one applies `code-scanning-selectors.yml` from the suite named `codeql/suite-helpers`. Looking that one up we find this [this file](https://github.com/github/codeql/blob/main/misc/suite-helpers/code-scanning-selectors.yml) which has the requirements.

Similarly, the formalization of the rules for the extended security suite can be found [here](https://github.com/github/codeql/blob/main/misc/suite-helpers/security-extended-selectors.yml).

I really hope we can bring this query to `@ precision high` so that it appears in code scanning by default.

Reply:
I did a run on LTGM and all instances seem to be true positives now, so setting `@precision high` seems reasonable.

Reply:
I agree, and that's great news as this query will have higher visibility. :+1: